soccer/gameplay/situational_play_selection.py

```diff
@@ -0,0 +1,742 @@
+    #
+    # The regions are defined in the update function
+    #
+    class FieldLoc(Enum):
+        DEFENDSIDE = 1
+        MIDFIELD = 2
+        ATTACKSIDE = 3
+
+    ##Enum for representing the possession of the ball
+    class BallPos(Enum):
+        OURBALL = 1
+        FREEBALL = 2
+        THEIRBALL = 3
+
+    ##An empty constructor
+    def __init__(self):
+        pass
+
+    ## Holds the current situation enum
+    currentSituation = Situation.NONE
+
+    ##Holds the current possession enum
+    currentPossession = BallPos.FREEBALL
+
+    ##Holds the ball location enum
+    ballLocation = FieldLoc.MIDFIELD
+
+    ##Represents if there is currently a pileup for the purposes of situation selection
+    currentPileup = False
+
+    ##Holds the last time that currentPileup was changed by the pileup update function 
+    pileupTime = time.time()
+
+    ##Variable used to ensure setup is done 
+    isSetup = False
+
+    ##Holds the game state object
+    gameState = None
+
+    ##Holds the system state object
+    systemState = None
+
+    ##Holds a list of all the robots
+    robotList: List[robocup.Robot] = list()
+
+    ##Holds a list of the currently visible robots
+    activeRobots: List[robocup.Robot] = list()
+
+    ##Function is run a single time to setup the class
+    #
+    # Gets the game state, system state, and context objects
+    #
+    # Builds the robotList for later use
+    #
+    def setupStates(self):
+        self.gameState = main.game_state()
+        self.systemState = main.system_state()
+        self.context = main.context()
+        for g in self.systemState.our_robots:
+            self.robotList.append(g)
+        for g in self.systemState.their_robots:
+            self.robotList.append(g)
+
+        self.updateRobotList()
+
+    ## Function that updates the activeRobots list
+    # 
+    # Updates the activerobots list every frame to include only the currently visible robots
+    #
+    def updateRobotList(self):
+        self.activeRobots.clear()
+        for g in self.robotList:
+            if (g.visible):
+                self.activeRobots.append(g)
+
+    ## Main update function
+    #   Is called every frame from main.py to update the situation information 
+    #   mostly calls all the other update functions
+    #   
+    #   Will also run the setup if isSetup is false and has functionality to get
+    #   the running time of the module
+    def updateAnalysis(self):
+        #startTime = time.time()
+        if (not self.isSetup):
+            self.setupStates()
+            self.isSetup = True
+        else:
+            self.updateRobotList()
+
+        self.updatePileup()
+        self.locationUpdate()
+        self.ballPossessionUpdate()
+        self.situationUpdate()
+
+        #print(abs(startTime - time.time()))
+
+        #Print the current situation in the corner of the soccer gui
+        self.context.debug_drawer.draw_text(self.currentSituation.name,
```

Comment:
It'd be nice at some point to throw this into the tree or something. That way we can see the state each frame since this is going to be very very important to have tuned correctly. There's a way to do it with normal plays, but I'm not sure about this specific case.